Isosurfacing_3/doc/Isosurfacing_3/Concepts/IsosurfacingDomain.h

```diff
@@ -0,0 +1,152 @@
+/*!
+\ingroup PkgIsosurfacing3Concepts
+\cgalConcept
+
+The concept `IsosurfacingDomain` describes the set of requirements to be
+fulfilled by any class used as input data for any isosurfacing algorithms.
+
+\cgalHasModel `CGAL::Isosurfacing::Cartesian_grid_domain`
+\cgalHasModel `CGAL::Isosurfacing::Implicit_domain`
+
+*/
+
+class IsosurfacingDomain {
+public:
+    /// \name Types
+    /// @{
+
+    /*!
+    Traits type model of \cgal %Kernel
+    */
+    typedef unspecified_type Traits;
+
+    /*!
+    The scalar type.
+    */
+    typedef unspecified_type FT;
+
+    /*!
+    The point type.
+    */
+    typedef unspecified_type Point;
+
+    /*!
+    The vector type.
+    */
+    typedef unspecified_type Vector;
+
+    /*!
+    A handle to identify a vertex.
+    */
+    typedef unspecified_type Vertex_handle;
+
+    /*!
+    A handle to identify an edge.
+    */
+    typedef unspecified_type Edge_handle;
+
+    /*!
+    A handle to identify a cell.
+    */
+    typedef unspecified_type Cell_handle;
+
+    /*!
+    A container for the two vertices of an edge.
+    */
+    typedef unspecified_type Edge_vertices;
+
+    /*!
+    A container for the cells incident to an edge.
+    */
+    typedef unspecified_type Cells_incident_to_edge;
+
+    /*!
+    A container for the vertices of a cell.
+    */
+    typedef unspecified_type Cell_vertices;
+
+    /*!
+    A container for the edges of a cell.
+    */
+    typedef unspecified_type Cell_edges;
+
+
+    /// @}
+
+    /// \name Operations
+    /// The following member functions must exist.
+    /// @{
+
+    /*!
+    Returns the position of vertex v in 3D space
+    */
+    Point position(const Vertex_handle& v) const;
+
+    /*!
+    Returns the value of vertex v
+    */
+    FT value(const Vertex_handle& v) const;
+
+    /*!
+    (Optional) Returns the gradient at the position p
+    */
+    Vector gradient(const Point& p) const;
+
+    /*!
+    Returns the two vertices incident to edge e
+    */
+    Edge_vertices edge_vertices(const Edge_handle& e) const;
+
+    /*!
+    Returns all voxels incident to edge e
+    */
+    Cells_incident_to_edge cells_incident_to_edge(const Edge_handle& e) const;
+
+    /*!
+    Returns all vertices of the cell c
+    */
+    Cell_vertices cell_vertices(const Cell_handle& c) const;
+
+    /*!
+    Returns all edges of the cell c
+    */
+    Cell_edges cell_edges(const Cell_handle& c) const;
+
+    /*!
```

Comment:
Use maybe a template parameter `ConcurrencyTag`?

Reply:
I don't think it is useful to parallelize over such a small collection. In most cases there will just be 12 edges and not much work done per edge.